Реализовал HTTP API

VectorID1 opened the pull request:
BaseHttpHandler новый клас с базовыми методами для обработки heldel`оф sendText sendNotFound sendHashInteraction
Создал новый класс DurationAdapter
Создал новый класс LocalDateTimeAdapter
Создал классы TaskHandler, EpicHandler, SubtaskHandler в них основная логима по обработке запросов от клиента. GET POST DELETE новые классы PriorityHandler и HistoryHandler обрабатывают запросы и возвращают списки задач История и Историяприоритетов Добавил 2 новых исключения
создал отдельтый csv файл для теста FileBackedTaskManagerTest Написал тесты всех новых классов, Добавляя удаляя и получая задачи из запросов клиента.

VectorID1 committed:
Реализовал HTTP API для
BaseHttpHandler новый клас с базовыми методами для обработки heldel`оф
sendText sendNotFound sendHashInteraction
Создал новый класс DurationAdapter
Создал новый класс LocalDateTimeAdapter
Создал классы TaskHandler, EpicHandler, SubtaskHandler в них основная логима по обработке запросов от клиента. GET POST DELETE
новые классы PriorityHandler и HistoryHandler обрабатывают запросы и возвращают списки задач История и Историяприоритетов
Добавил 2 новых исключения
создал отдельтый csv файл для теста FileBackedTaskManagerTest
Написал тесты всех новых классов, Добавляя удаляя и получая задачи из запросов клиента.

vlolad commented on src/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,62 @@
+    public void start() throws IOException {
+        server = HttpServer.create(new InetSocketAddress(PORT), 0);
+
+        server.createContext("/tasks", new TaskHandler(taskManager, gson));
```
Лучше инициализацию контекста тоже в конструктор перенести

vlolad commented on src/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,62 @@
+        this.gson = new GsonBuilder()
+                .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeAdapter())
+                .registerTypeAdapter(Long.class, new DurationAdapter())
+                .registerTypeAdapter(long.class, new DurationAdapter())
```
Дважды один и тот же адаптер регистрируется

vlolad commented on src/http/handler/EpicsHandler.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```
В таком варианте можно попробовать в BaseHttpHandler вынести, достаточно универсально получилось. Если где-то, допустим, POST не нужен - возвращать NOT_IMPLEMENTED

vlolad replied:
Но вообще всё отлично, это на улучшение комментарий

VectorID1 replied:
Я пока трогать не буду, ты имеешь ввиду что метод общий для всех хендлеров сделать?  Разве не может быть дальше так что в каком то классе не понадобиться какой то эндпоинт? и куда его тогда девать? Может оставить? Как говорится Работает не трогай =)))

vlolad commented on src/http/handler/adapter/DurationAdapter.java:
```diff
@@ -0,0 +1,29 @@
+public class DurationAdapter extends TypeAdapter<Long> {
+
+    @Override
+    public void write(JsonWriter out, Long duration) throws IOException {
```
nullValue лучше не использовать. Если duration пустой, то записываем 0L

vlolad commented on src/http/handler/adapter/DurationAdapter.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public Long read(JsonReader in) throws IOException {
+        if (in.peek() == JsonToken.NULL) {
```
Можно через Duration.ofMinutes(Long.parseLong(/*строка*/)) и ловить NumberFormatException, но в целом ок

vlolad commented on src/http/handler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```
Тут не надо handle вообще добавлять, сделай класс BaseHttpHandler абстрактным
Конкретная реализация в наследниках будет идти

vlolad reviewed (changes requested):
Привет, в целом всё отлично, только совсем небольшие ошибки на исправление

И будем закрывать проект с:

VectorID1 committed:
Поправил Адаптер DurationAdapter
Перенес инициализацию контекста при создании сервера в конструктор HttpTaskServer
Сделал класс BaseHttpHandler абстрактным и удалил из него hendle

vlolad reviewed (approved):
Привет, всё поправлено, принято (￣^￣)ゞ

Теперь переходим к изучению фреймворка, больше такой низкоуровневой настройки сервера вручную (почти) не будет
